sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java

```diff
@@ -147,6 +150,42 @@
  * Read#withCustomRateLimitPolicy(RateLimitPolicyFactory)}. This requires implementing {@link
  * RateLimitPolicy} with a corresponding {@link RateLimitPolicyFactory}.
  *
+ * <h4>Enhanced Fan-Out</h4>
+ *
+ * Kinesis IO supports Consumers with Dedicated Throughput (Enhanced Fan-Out, EFO). This type of
+ * consumer doesn't have to contend with other consumers that are receiving data from the stream.
+ *
+ * <p>More details can be found here: <a
+ * href="https://docs.aws.amazon.com/streams/latest/dev/enhanced-consumers.html">Consumers with
+ * Dedicated Throughput</a>
+ *
+ * <p>EFO can be enabled for one or more {@link KinesisIO.Read} instances via pipeline options:
```

Comment:
This brings 2 issues:
- Having 2 ways of configuring the same thing will be confusing, I think. And having those 2 ways behaving differently is even more confusing: when you drop `withConsumerArn()` from read spec, it actually doesn't go away and stays in the savepoints forever.
- Imagine a scenario: I start from LATEST with EFO, then EFO becomes not necessary. How do I disable EFO? Let's say I disable via NOT providing PipelineOptions. Then, it's not a default anymore.

I know that it's a bit confusing too - NOT to have `withConsumerArn()`, but the real issue I see is that framework actually assumes `Read` instances not being effectively editable.

Reply:
yes, I agree that this is by far not optimal ... I fear without adding it to the read spec too many users won't even notice that there's EFO support and configuration of IOs using pipeline options is not common (actually, even discouraged if not absolutely necessary for exactly that reason).

So I guess reasonable semantics for the configuration would be:
- If you provide a consumerArn via the read spec, EFO will ALWAYS be turned on (that should be the default use case, either have it on or off).
- To temporarily turn it on (e.g. to backfill), one can temporarily provide consumer arns via pipeline options and remove them again.

Reply:
> If you provide a consumerArn via the read spec, EFO will ALWAYS be turned on (that should be the default use case, either have it on or off).

Still, it's very counter-intuitive that dropping `withConsumerArn()` does not disable EFO. And if the consumer is de-registered by a mistake, and then a new one is registered, the user will need to provide Pipeline Options.

I will meditate a bit more on this, your points are sound, too 👍

Reply:
> I fear without adding it to the read spec too many users won't even notice that there's EFO support

How about adding something like `witEnhancedFanOutEnabled(boolean)`?

- If set - the source will try to lookup a consumer ARN from PipelineOptions
    - If consumer ARN not found - a warning is thrown and non-EFO consumer starts
    - If consumer ARN found - EFO consumer starts
- If not set:
    - If consumer ARN not found - start non-EFO consumer without a warning
    - If consumer ARN found - EFO consumer starts

When defined like that, it will serve only to "document" pipeline code, but enabling / disabling / changing consumer ARN will be done solely in PipelineOptions.

What do you think?

Reply:
@psolomin Keep in mind, there's still other runners that don't share the behavior of Flink. With Dataflow I'm 99% sure you can change the configuration when updating / resuming a streaming pipeline (https://cloud.google.com/dataflow/docs/guides/updating-a-pipeline). Though, happy to confirm that with some Googlers if you prefer. Honestly, I'd be rather reluctant to add a non-functional configuration such as `withEnhancedFanOutEnabled()` that then fully relies on configuration via pipeline options.
I feel to comply with Beam standards it should be possible to configure this feature alone using the read spec (even though that is very limited when used with Flink). 

Regarding the inconsistency discussed above (enabled via read spec, but how to disable then), it wouldn't be hard to use the pipeline options mechanism as well using a null mapping:

```
--kinesisIOConsumerArns={"stream1":null}
```

Just to get a 3rd opinion on his, @aromanenko-dev any thoughts from your side? 

One use case we discussed is to temporarily enabled EFO consumers to initially backfill a pipeline with max. throughput, but then disable EFO again once the pipeline caught up to save on costs.

Context to this discussion is the question how to enable Enhanced Fanout Consumers for Kinesis to support the above use case as well. Problem is that the read specification is serialized together with the source in Flink savepoints. So when resuming the pipeline any changes to the read configuration won't be noticed - and EFO can't be disabled. The only way to do this (for Flink) is to configure EFO consumers via pipeline options.

Reply:
> Dataflow

To me, introducing a config which has different semantics on different runners is even more questionable. But it sounds like we need an opinion of someone who could say if this is a norm for other IOs.

Reply:
> To me, introducing a config which has different semantics on different runners is even more questionable. But it sounds like we need an opinion of someone who could say if this is a norm for other IOs.

The norm is configuration of IOs is always done via the read / write spec, see the [IO standard docs](https://beam.apache.org/documentation/io/io-standards/#all-sdks):

> An I/O should rarely rely on a PipelineOptions subclass to tune internal parameters. If neccesary, a connector-related pipeline options class should ...

 The pipeline option is a just an exceptional workaround to support a valid use case on some runner (namely the Flink runner). 

Maybe we should postpone support for this use case all together. It could be a follow up in another PR.

Reply:
> To me, introducing a config which has different semantics on different runners is even more questionable.

The semantics of the read spec are not anything that is specific to this IO. This is the status quo and applies to everything. But yes, it is unfortunate that semantics are different for the various runners.

Reply:
> Maybe we should postpone support for this use case all together. It could be a follow up in another PR

This will actually complicate my development iterations which involve tests with Flink runner :(

And since we have `--kinesisIOConsumerArns={"stream1":null}` work-around, I would give it a try in this PR and see if the end result and documentation is not too confusing.

Reply:
@mosche it looks like in Beam 2.46.0 we have an issue with `Read`: original class and generated class IDs don't match.

```
$ serialver -classpath "target/dependency/*" \
  org.apache.beam.sdk.io.aws2.kinesis.KinesisIO.Read
org.apache.beam.sdk.io.aws2.kinesis.KinesisIO.Read:    private static final long serialVersionUID = 1118639910791038161L;

$ serialver -classpath "target/dependency/*" \
  org.apache.beam.sdk.io.aws2.kinesis.AutoValue_KinesisIO_Read
org.apache.beam.sdk.io.aws2.kinesis.AutoValue_KinesisIO_Read:    private static final long serialVersionUID = 410184916277145716L;
```

If I set `serialVersionUID` in `Read`, `AutoValue_KinesisIO_Read` gets the same one - still not matching the Beam 2.46.0

Any tips? If there is no way to have IDs in `Read` and `AutoValue_KinesisIO_Read` both hard-coded, what would be a way forward?

- Keep `Read` unchanged in this PR, create an issue and have compatibility-breaking change of it in other release
- Break compatibility in this PR, adding a note to the changelog accordingly

Reply:
Done. Basically, I did this:

- Broke serialisation compatibility with Beam 2.46.0 for `KinesisIO.Read` and `KinesisSource` for Flink runner
- Maintained serialisation compatibility of `KinesisReaderCheckpoint` with Beam 2.46.0 - this one, I think, we should keep - other runners may be not persisting `KinesisSource` but I am sure they persist the checkpoint.